sign/bls2/bls2.go

```diff
@@ -0,0 +1,150 @@
+// Package bls2 implements the robust BLS scheme that prevents rogue public-key
+// to be used to forge signatures
+// See the paper: https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html
+package bls2
+
+import (
+	"crypto/cipher"
+	"errors"
+	"math/big"
+
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/group/mod"
+	"go.dedis.ch/kyber/v3/pairing"
+	"go.dedis.ch/kyber/v3/sign/bls"
+	"golang.org/x/crypto/blake2b"
+)
+
+var modulus128 = new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 128), big.NewInt(1))
+
+// For the choice of H, we're mostly worried about the second preimage attack. In
+// other words, find H(m) == H(m')
+// We also use the entire roster so that the coefficient will vary for the same
+// public key used in different roster
+func hashPointToR(point []byte, roster [][]byte) (kyber.Scalar, error) {
+	h, err := blake2b.New(16, nil)
```

Comment:
Out of curiosity: Is there a specific reason why BLAKE2 instead of SHA{256,512} is used here? The current IETF draft standard seems to suggest usage of SHA512: https://www.ietf.org/id/draft-boneh-bls-signature-00.txt (but it also uses the curve BLS12-381)

Reply:
I followed the Stanford paper where they're saying the hash function should map to the set [1; 2^128] so I took BLAKE2 because of the length of the digest. (and supposedly better than MD5)

Reply:
According to the BLAKE2b documentation the output should be at least 32 bytes, i.e., 256 bits, if BLAKE2b is used as a hash function (see `New` in https://godoc.org/golang.org/x/crypto/blake2b) to get to the minimum security level of 128 bit.

It's not clear to me how this maintains a 128-bit security level, as BLAKE2b (or any hash function fwiw) achieves only 64 bits of security for a 128-bit output (due to the birthday bound).

Reply:
My reasoning behing this, and pls correct me if I'm wrong, is that we want to be safe against a 2nd preimage attack because the attacker needs to find `t*` that is equal to `t` where `t` is the coefficient of the honest public key. Isn't the birthday attack only right for the collision resistance ?

In any case, finding a 2nd preimage doesn't mean you can forge the public key because you still need to find the private key associated with it: `g^alpha * (1/pk) == 2nd preimage` meaning you still need to solve the DL problem.